File: scripts/check-site.mjs
import fs from "node:fs";
import path from "node:path";

const root = process.cwd();
const siteOrigin = "https://foshchii.github.io";
const ignoredDirs = new Set([".git", "node_modules", "tmp"]);
const ignoredSchemes = /^(?:https?:|mailto:|tel:|sms:|webcal:|data:|javascript:)/i;
const checks = [];
const failures = [];

function walk(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const files = [];

  for (const entry of entries) {
    if (ignoredDirs.has(entry.name)) continue;
    const target = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...walk(target));
    } else {
      files.push(target);
    }
  }

  return files;
}

const htmlFiles = walk(root)
  .filter((file) => file.endsWith(".html"))
  .map((file) => path.relative(root, file))
  .sort();

function fileExists(target) {
  try {
    return fs.statSync(target).isFile();
  } catch {
    return false;
  }
}

function directoryExists(target) {
  try {
    return fs.statSync(target).isDirectory();
  } catch {
    return false;
  }
}

function stripQueryAndHash(value) {
  return value.split("#")[0].split("?")[0];
}

function decodePath(value) {
  try {
    return decodeURI(value);
  } catch {
    return value;
  }
}

function resolveInternalReference(fromFile, value) {
  if (!value || value.startsWith("#")) return null;

  const cleanValue = value.trim();
  if (!cleanValue || cleanValue.startsWith("//")) return null;

  if (cleanValue.startsWith(siteOrigin)) {
    return path.join(root, new URL(cleanValue).pathname);
  }

  if (ignoredSchemes.test(cleanValue)) return null;

  const clean = decodePath(stripQueryAndHash(cleanValue)).trim();
  if (!clean) return null;

  if (clean.startsWith("/")) {
    return path.join(root, clean);
  }

  return path.resolve(root, path.dirname(fromFile), clean);
}

function collectReferences(html) {
  const refs = [];
  const patterns = [
    /\b(?:href|src|poster|action)=["']([^"']+)["']/gi,
    /\bsrcset=["']([^"']+)["']/gi,
    /\bcontent=["'](https:\/\/foshchii\.github\.io\/[^"']+)["']/gi
  ];

  for (const pattern of patterns) {
    let match;
    while ((match = pattern.exec(html))) {
      if (pattern.source.includes("srcset")) {
        for (const candidate of match[1].split(",")) {
          const [url] = candidate.trim().split(/\s+/);
          refs.push(url);
        }
      } else {
        refs.push(match[1]);
      }
    }
  }

  return refs;
}

function getAttr(html, selector) {
  const match = html.match(selector);
  return match ? match[1] : "";
}

function canonicalToPath(canonical) {
  if (!canonical.startsWith(siteOrigin)) return null;
  const url = new URL(canonical);
  if (url.pathname === "/") return "index.html";
  if (url.pathname.endsWith("/")) return path.join(url.pathname.slice(1), "index.html");
  return url.pathname.slice(1);
}

const sitemap = fs.readFileSync(path.join(root, "sitemap.xml"), "utf8");

for (const file of htmlFiles) {
  const absolute = path.join(root, file);
  const html = fs.readFileSync(absolute, "utf8");
  const isNoindex = /<meta\s+name=["']robots["']\s+content=["'][^"']*noindex/i.test(html);
  checks.push(`${file}: readable`);

  for (const ref of collectReferences(html)) {
    const target = resolveInternalReference(file, ref);
    if (!target) continue;

    const clean = stripQueryAndHash(ref);
    const pointsToRoot = clean === "/" || clean === ".";
    const ok = pointsToRoot
      ? directoryExists(target)
      : fileExists(target) || directoryExists(target) || fileExists(`${target}.html`);

    checks.push(`${file}: ${ref}`);
    if (!ok) {
      failures.push(`${file} references missing path: ${ref}`);
    }
  }

  const h1Count = (html.match(/<h1\b/gi) || []).length;
  checks.push(`${file}: h1`);
  if (h1Count !== 1) failures.push(`${file} must contain exactly one h1; found ${h1Count}`);

  const title = getAttr(html, /<title>([^<]+)<\/title>/i).trim();
  checks.push(`${file}: title`);
  if (!title) failures.push(`${file} is missing a title`);

  const canonical = getAttr(html, /<link\s+rel=["']canonical["']\s+href=["']([^"']+)["']/i).trim();
  if (!isNoindex) {
    checks.push(`${file}: canonical`);
    if (!canonical) failures.push(`${file} is missing a canonical URL`);

    const description = getAttr(html, /<meta\s+name=["']description["']\s+content=["']([^"']+)["']/i).trim();
    checks.push(`${file}: description`);
    if (!description) failures.push(`${file} is missing a meta description`);

    const expectedPath = canonical ? canonicalToPath(canonical) : null;
    if (expectedPath && path.normalize(expectedPath) !== path.normalize(file)) {
      failures.push(`${file} canonical points to ${expectedPath}`);
    }

    if (canonical && !sitemap.includes(`<loc>${canonical}</loc>`)) {
      failures.push(`${file} canonical is missing from sitemap: ${canonical}`);
    }

    const ogImage = getAttr(html, /<meta\s+property=["']og:image["']\s+content=["']([^"']+)["']/i).trim();
    checks.push(`${file}: og:image`);
    if (!ogImage) {
      failures.push(`${file} is missing og:image`);
    } else {
      const target = resolveInternalReference(file, ogImage);
      if (target && !fileExists(target)) {
        failures.push(`${file} og:image does not exist locally: ${ogImage}`);
      }
    }

    checks.push(`${file}: feed link`);
    if (!html.includes('type="application/rss+xml"') || !html.includes(`${siteOrigin}/feed.xml`)) {
      failures.push(`${file} is missing the RSS feed link`);
    }
  }

  const jsonLdBlocks = [...html.matchAll(/<script\s+type=["']application\/ld\+json["']>([\s\S]*?)<\/script>/gi)];
  for (const [index, block] of jsonLdBlocks.entries()) {
    checks.push(`${file}: JSON-LD ${index + 1}`);
    try {
      JSON.parse(block[1]);
    } catch (error) {
      failures.push(`${file} has invalid JSON-LD block ${index + 1}: ${error.message}`);
    }
  }
}

const requiredFiles = [
  ".nojekyll",
  "feed.xml",
  "robots.txt",
  "sitemap.xml",
  "llms.txt",
  "assets/Sviatoslav-Foshchii-CV.pdf",
  "assets/css/styles.css",
  "assets/js/main.js",
  "assets/js/booking-widget.js"
];

for (const file of requiredFiles) {
  checks.push(file);
  if (!fileExists(path.join(root, file))) {
    failures.push(`Required file is missing: ${file}`);
  }
}

const feed = fs.readFileSync(path.join(root, "feed.xml"), "utf8");
checks.push("feed.xml: rss");
if (!/<rss\b[^>]*version=["']2\.0["']/i.test(feed) || !feed.includes(`${siteOrigin}/insights.html`)) {
  failures.push("feed.xml is not a valid RSS 2.0 article feed");
}

if (failures.length) {
  console.error("Site check failed:");
  for (const failure of failures) console.error(`- ${failure}`);
  process.exit(1);
}

console.log(`Site check passed: ${checks.length} checks across ${htmlFiles.length} HTML files.`);

File: .github/scripts/check_site.py
#!/usr/bin/env python3
"""Pre-deploy checks for the site. Standard library only; needs Node for the
JavaScript syntax check.

    python3 .github/scripts/check_site.py

Catches the mistakes that are easy to make when adding or editing a page and
invisible until someone hits them on the live site:
  - links, images, scripts and stylesheets that point at a missing file, and
    #anchors that point at a missing id
  - structured data (JSON-LD) that is not valid JSON, which search engines
    silently ignore
  - a canonical or og:url that does not match the page, as happens when a new
    page is copied from an old one
  - pages missing from sitemap.xml, or sitemap and llms.txt entries that point
    at nothing
  - feed.xml out of date with the articles
  - the booking backend's VERSION out of date with its code
  - JavaScript syntax errors
"""

import json
import os
import re
import shutil
import subprocess
import sys
import xml.etree.ElementTree as ET
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import unquote, urlsplit

sys.path.insert(0, str(Path(__file__).resolve().parent))
import build_feed  # noqa: E402
import stamp_backend_version  # noqa: E402

ROOT = build_feed.ROOT
SITE = build_feed.SITE
NOT_INDEXED = {"404.html"}
EXTERNAL = re.compile(r"^([a-z][a-z0-9+.-]*:|//)", re.I)

errors = []


def fail(file, line, msg):
    errors.append((file, line, msg))


class Page(HTMLParser):
    def __init__(self, name):
        super().__init__()
        self.name = name
        self.ids = set()
        self.refs = []       # (line, url)
        self.jsonld = []     # (line, text)
        self.canonical = []
        self.og_url = []
        self._script = None  # (line, type, [chunks]) while inside <script>

    def handle_starttag(self, tag, attrs):
        line = self.getpos()[0]
        a = dict(attrs)
        if a.get("id"):
            self.ids.add(a["id"])
        for attr in ("href", "src"):
            if a.get(attr):
                self.refs.append((line, a[attr]))
        if a.get("srcset"):
            for candidate in a["srcset"].split(","):
                if candidate.strip():
                    self.refs.append((line, candidate.split()[0]))
        if tag == "link" and a.get("rel") == "canonical":
            self.canonical.append(a.get("href"))
        if tag == "meta" and a.get("property") == "og:url":
            self.og_url.append(a.get("content"))
        if tag == "script":
            self._script = (line, a.get("type", ""), [])

    def handle_data(self, data):
        if self._script:
            self._script[2].append(data)

    def handle_endtag(self, tag):
        if tag != "script" or not self._script:
            return
        line, kind, chunks = self._script
        text = "".join(chunks)
        if kind == "application/ld+json":
            self.jsonld.append((line, text))
        else:
            for m in re.finditer(r"""import\(\s*['"]([^'"]+)['"]\s*\)""", text):
                self.refs.append((line + text[: m.start()].count("\n"), m.group(1)))
        self._script = None


def parse(name):
    p = Page(name)
    p.feed((ROOT / name).read_text(encoding="utf-8"))
    p.close()
    return p


def url_for(page):
    return SITE if page == "index.html" else SITE + page


def page_for(url):
    """Map an absolute site URL to a file path relative to ROOT."""
    rel = unquote(urlsplit(url).path).lstrip("/")
    return rel + "index.html" if rel == "" or rel.endswith("/") else rel


def check_references(pages):
    count = 0
    for name, page in pages.items():
        for line, url in page.refs:
            if EXTERNAL.match(url):
                continue
            count += 1
            parts = urlsplit(url)
            path = unquote(parts.path)
            if not path:
                target = name
            elif path.startswith("/"):
                target = page_for(path)
            else:
                target = os.path.normpath(os.path.join(os.path.dirname(name), path))
                if path.endswith("/"):
                    target += "/index.html"
            if not (ROOT / target).is_file():
                fail(name, line, f"'{url}' points at {target}, which does not exist")
                continue
            if parts.fragment and target.endswith(".html"):
                ids = pages[target].ids if target in pages else parse(target).ids
                if parts.fragment not in ids:
                    fail(name, line, f"'{url}': {target} has no element with id=\"{parts.fragment}\"")
    return count


def check_jsonld(pages):
    count = 0
    for name, page in pages.items():
        for line, text in page.jsonld:
            count += 1
            try:
                json.loads(text)
            except json.JSONDecodeError as e:
                fail(name, line + e.lineno - 1, f"JSON-LD is not valid JSON: {e.msg}")
    return count


def check_canonical(pages):
    for name, page in pages.items():
        if name in NOT_INDEXED:
            continue
        want = url_for(name)
        for label, got in (("canonical", page.canonical), ("og:url", page.og_url)):
            if got != [want]:
                found = ", ".join(map(str, got)) or "none"
                fail(name, 1, f"{label} should be {want} (found: {found})")


def check_sitemap(pages):
    try:
        root = ET.parse(ROOT / "sitemap.xml").getroot()
    except ET.ParseError as e:
        fail("sitemap.xml", e.position[0], f"not well-formed XML: {e}")
        return
    ns = {"s": "http://www.sitemaps.org/schemas/sitemap/0.9"}
    listed = []
    for loc in root.findall("s:url/s:loc", ns):
        url = (loc.text or "").strip()
        rel = page_for(url)
        if not url.startswith(SITE):
            fail("sitemap.xml", 1, f"{url} is not on {SITE}")
        elif rel not in pages:
            fail("sitemap.xml", 1, f"{url} points at {rel}, which does not exist")
        elif rel in listed:
            fail("sitemap.xml", 1, f"{url} is listed more than once")
        listed.append(rel)
    for name in sorted(set(pages) - NOT_INDEXED - set(listed)):
        fail("sitemap.xml", 1, f"{name} is not listed; add {url_for(name)}")


def check_llms():
    text = (ROOT / "llms.txt").read_text(encoding="utf-8")
    for n, line in enumerate(text.splitlines(), 1):
        for url in re.findall(re.escape(SITE) + r"[^\s)>\]]*", line):
            if not (ROOT / page_for(url)).is_file():
                fail("llms.txt", n, f"{url} points at {page_for(url)}, which does not exist")


def check_feed():
    try:
        want = build_feed.render()
    except (ValueError, KeyError, json.JSONDecodeError) as e:
        fail("feed.xml", 1, f"cannot build the feed: {e}")
        return
    path = ROOT / build_feed.FEED
    if not path.is_file() or path.read_text(encoding="utf-8") != want:
        fail("feed.xml", 1, "out of date with the articles; run: python3 .github/scripts/build_feed.py")


def check_backend_version():
    script = stamp_backend_version.SCRIPT
    try:
        text = stamp_backend_version.read()
        m = stamp_backend_version.find(text)
    except (OSError, ValueError) as e:
        fail(script, 1, str(e))
        return
    if m.group(1) != stamp_backend_version.version(text):
        fail(script, text[: m.start()].count("\n") + 1,
             "VERSION is out of date with the code; run: python3 .github/scripts/stamp_backend_version.py")


def check_js():
    node = shutil.which("node")
    if not node:
        print("note: node not found, skipping the JavaScript syntax check", file=sys.stderr)
        return 0
    files = sorted(ROOT.glob("*.js")) + sorted(ROOT.glob("assets/js/*.js"))
    for f in files:
        r = subprocess.run([node, "--check", str(f)], capture_output=True, text=True)
        if r.returncode:
            rel = f.relative_to(ROOT).as_posix()
            m = re.search(re.escape(str(f)) + r":(\d+)", r.stderr)
            detail = next((l for l in r.stderr.splitlines() if "Error" in l), r.stderr.strip())
            fail(rel, int(m.group(1)) if m else 1, f"JavaScript syntax error: {detail}")
    return len(files)


def main():
    pages = {p.name: parse(p.name) for p in sorted(ROOT.glob("*.html"))}
    refs = check_references(pages)
    blocks = check_jsonld(pages)
    check_canonical(pages)
    check_sitemap(pages)
    check_llms()
    check_feed()
    check_backend_version()
    scripts = check_js()

    annotate = os.environ.get("GITHUB_ACTIONS") == "true"
    for file, line, msg in errors:
        print(f"{file}:{line}: {msg}")
        if annotate:
            print(f"::error file={file},line={line}::{msg}")
    if errors:
        print(f"\n{len(errors)} problem(s) found.")
        return 1
    print(f"ok: {len(pages)} pages, {refs} internal links, {blocks} JSON-LD blocks, "
          f"{scripts} scripts, sitemap, llms.txt, feed and backend version all check out.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
